### PR TITLE
Module8 branch

### PR DESCRIPTION
**Pull request by Linases**
Working on Pages and Tests.
Done and works (if no Alert is present for a discount genius): Login and Language Change scenarios.
I tried to apply dismiss this alert in Base test, but it does not always find this element. 

Started other scenarios, but encountering some issues with finding elements or checking assertions, so maybe I will divide scenarios in smaller methods.

### CODE REVIEW

File: Utilities/WebDriverExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            return wait.Until(ExpectedConditions.ElementIsVisible(locator));
+        }
+
+        public static IList<IWebElement> WaitForElementsVisible(this IWebDriver driver, By locator, int timeoutInSeconds = 30)
+        {
+            var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(timeoutInSeconds));
+            return wait.Until(ExpectedConditions.PresenceOfAllElementsLocatedBy(locator));
+        }
+
+        public static IWebElement WaitForElementClicable(this IWebDriver driver, By locator, int timeoutInSeconds = 20)
```
Comment (Alina150601) [inline Utilities/WebDriverExtensions.cs]: WaitForElement**Is**Clicable

Comment (Alina150601) [reply Utilities/WebDriverExtensions.cs]: check other methods here too

File: Booking_Tests_Waits/SearchTests.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            Assert.That(destination.Contains($"{_expectedDestination}"), Is.True, $" {_expectedDestination} is not entered destination");
+        }
+
+        private void EnterCheckInCheckOutDays()
+        {
+            _datePickerPage.OpenCalendar();
+            _datePickerPage.SelectDate(_checkInDate);
+            _datePickerPage.SelectDate(_checkOutDate);
+            var checkIn = _datePickerPage.GetSelectedCheckInDay();
+            DateTime parsedDate = DateTime.ParseExact(checkIn, "ddd, MMM dd", System.Globalization.CultureInfo.InvariantCulture);
```
Comment (Alina150601) [inline Booking_Tests_Waits/SearchTests.cs]: var

File: Booking_Tests_Waits/SearchTests.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        {
+            _searchPage.ClickGuestInput();
+            _searchPage.SelectAdultsNr(_expectedAdultNr);
+            _searchPage.SelectChildrenNr(_expectedChildrenNr);
+            _searchPage.SelectRoomsNr(_expectedRoomsNr);
+            _searchPage.PressDone();
+
+            var actualGuestsNr = _searchPage.GetGuestsNrValue();
+            var expectedGuestsNr = $"{_expectedAdultNr}{_expectedChildrenNr}{_expectedRoomsNr}";
+            bool containsAllNumbers = expectedGuestsNr.All(c => actualGuestsNr.Contains(c));
```
Comment (Alina150601) [inline Booking_Tests_Waits/SearchTests.cs]: does the assert fail, if number of children, adults and rooms will be mixed and will be not in right order?

File: BookingPages/AirportTaxiPage.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        public void SelectTaxi()
+        {
+            var expensivePrice = _driver.GetWaitForElementsVisible(SearchResultsList);
+            var lastTaxi = expensivePrice.LastOrDefault(x => x.Displayed);
+            lastTaxi.Click();
+        }
+
+        public void ClickContinueButton() => ContinueButton.Click();
+
+        public bool SummaryDisplayed() => ItinerarySummary.Displayed;
```
Comment (Alina150601) [inline BookingPages/AirportTaxiPage.cs]: bool methods should start with Is...

File: BookingPages/AirportTaxiPage.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+        public bool isDisplayedList()
```
Comment (Alina150601) [inline BookingPages/AirportTaxiPage.cs]: uppercase, and IsListDisplayed()

File: BookingPages/AttractionsPage.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        public void EnterDestination(string destination) => SearchField.SendKeys(destination);
+
+        public string GetDestination() => SearchField.GetAttribute("value");
+
+        public void SelectAutocompleteOption()
+        {
+            var destination = GetDestination();
+            IList<IWebElement> autocomplete = _driver.WaitForElementsVisible(AutocompleteList);
```
Comment (Alina150601) [inline BookingPages/AttractionsPage.cs]: var, check evetywhere

File: BookingPages/LanguagePage.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
+﻿using OpenQA.Selenium;
+using Utilities;
+
+namespace BookingPages
+{
+    public class LanguagePage
+    {
+        private readonly IWebDriver _driver;
+        private By LanguageElement => By.XPath("//*[text()='Nederlands']");
```
Comment (Alina150601) [inline BookingPages/LanguagePage.cs]: does always text is Nederlands? what if in another test will be another country?

File: BookingPages/LoginPage.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        public void EnterEmail(string email) => EmailField.SendKeys(email);
+
+        public string GetEmail() => EmailField.GetAttribute("value");
+
+        public void ClickContinueButton() => ContinueWithEmailButton.Click();
+
+        public bool IsPasswordFieldVisible()
+        {
+            var Password = _driver.WaitForElementVisible(By.XPath("//*[@placeholder='Enter a password']"));
+            return Password.Displayed;
```
Comment (Alina150601) [inline BookingPages/LoginPage.cs]: usually we create a variable before returning in, for easier debugging 
check the same below please

File: BookingPages/SearchPage.cs [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+
+        public void Select5Stars()
+        {
+            var stars = _driver.GetWait().Until(ExpectedConditions.ElementToBeClickable(Checkbox5stars));
+            stars.Click();
+        }
+
+        public string GetSearchResults()
+        {
+            _driver.GetWait().Until(ExpectedConditions.TextToBePresentInElementLocated(SearchResults, GetFiveStartsHotelsCeckboxValue()));
```
Comment (Alina150601) [inline BookingPages/SearchPage.cs:162]: blank line

File: BookingPages/LanguagePage.cs (hunk 7 above)
Comment (Linases) [reply BookingPages/LanguagePage.cs]: Changed

File: BookingPages/LoginPage.cs (hunk 8 above)
Comment (Linases) [reply BookingPages/LoginPage.cs]: changed

File: BookingPages/AttractionsPage.cs (hunk 6 above)
Comment (Linases) [reply BookingPages/AttractionsPage.cs]: checked

File: BookingPages/AirportTaxiPage.cs (hunk 5 above)
Comment (Linases) [reply BookingPages/AirportTaxiPage.cs]: Changed

File: BookingPages/AirportTaxiPage.cs (hunk 4 above)
Comment (Linases) [reply BookingPages/AirportTaxiPage.cs]: Changed

File: Booking_Tests_Waits/SearchTests.cs (hunk 3 above)
Comment (Linases) [reply Booking_Tests_Waits/SearchTests.cs]: does not

File: Booking_Tests_Waits/SearchTests.cs (hunk 2 above)
Comment (Linases) [reply Booking_Tests_Waits/SearchTests.cs]: fixed

File: Utilities/WebDriverExtensions.cs (hunk 1 above)
Comment (Linases) [reply Utilities/WebDriverExtensions.cs]: ok, changed everywhere

File: Booking_Tests_Waits/SearchTests.cs (hunk 3 above)
Comment (Linases) [reply Booking_Tests_Waits/SearchTests.cs]: changed

**Pull request merged**